pkg/modeling/modeling.go

```diff
@@ -0,0 +1,403 @@
+		},
+		map[ResourceName]resource.Quantity{
+			ResourceCPU:    *resource.NewMilliQuantity(4, resource.DecimalSI),
+			ResourceMemory: *resource.NewQuantity(1024*4, resource.DecimalSI),
+		},
+		map[ResourceName]resource.Quantity{
+			ResourceCPU:    *resource.NewMilliQuantity(8, resource.DecimalSI),
+			ResourceMemory: *resource.NewQuantity(1024*8, resource.DecimalSI),
+		},
+		map[ResourceName]resource.Quantity{
+			ResourceCPU:    *resource.NewMilliQuantity(16, resource.DecimalSI),
+			ResourceMemory: *resource.NewQuantity(1024*16, resource.DecimalSI),
+		},
+		map[ResourceName]resource.Quantity{
+			ResourceCPU:    *resource.NewMilliQuantity(32, resource.DecimalSI),
+			ResourceMemory: *resource.NewQuantity(1024*32, resource.DecimalSI),
+		},
+		map[ResourceName]resource.Quantity{
+			ResourceCPU:    *resource.NewMilliQuantity(64, resource.DecimalSI),
+			ResourceMemory: *resource.NewQuantity(1024*64, resource.DecimalSI),
+		},
+		map[ResourceName]resource.Quantity{
+			ResourceCPU:    *resource.NewMilliQuantity(128, resource.DecimalSI),
+			ResourceMemory: *resource.NewQuantity(1024*128, resource.DecimalSI),
+		},
+		map[ResourceName]resource.Quantity{
+			ResourceCPU:    *resource.NewMilliQuantity(256, resource.DecimalSI),
+			ResourceMemory: *resource.NewQuantity(1024*256, resource.DecimalSI),
+		},
+		map[ResourceName]resource.Quantity{
+			ResourceCPU:    *resource.NewMilliQuantity(512, resource.DecimalSI),
+			ResourceMemory: *resource.NewQuantity(1024*512, resource.DecimalSI),
+		},
+	}
+)
+
+type modelingSummary []resourceModels
+
+// resourceModels records the number of each allocatable resource models.
+// models is a pointer, it points to the address of the first model
+// You don't need to care about the data structure behind the first model.
+type resourceModels struct {
+	// quantity is the total number of each allocatable resource models
+	// +required
+	Quantity int
+
+	// when the number of node is less than or equal to six, it will be sorted by linkedlist,
+	// when the number of node is more than six, it will be sorted by red-black tree.
+
+	// when the data structure is linkedlist,
+	// each item will store clusterResourceNode.
+	// +required
+	linkedlist *list.List
+
+	// when the data structure is redblack tree,
+	// each item will store a key-value pair,
+	// key is ResourceList, the value is quantity of this ResourceList
+	// +optional
+	redblackTree *rbt.Tree
+}
+
+// clusterResourceNode represents the each raw resource entity without modeling.
+type clusterResourceNode struct {
+	// quantity is the the number of this node
+	// Only when the resourceLists are exactly the same can they be counted as the same node.
+	// +required
+	quantity int
+
+	// resourceList records the resource list of this node.
+	// It maybe contain cpu, mrmory, gpu...
+	// User can specify which parameters need to be included before the cluster starts
+	// +required
+	resourceList ResourceList
```

Comment:
corev1.ResourceList

Reply:
Here need to use a custom ResourceList data structure. Because ResourceName is an enum maintained by API. The key of ResourceList is ResourceName, which is a custom data structure.

Reply:
Then we need to provide a method to convert corev1.resourcelist to Resourcelist.